docs_src/design/adr/core/0019-EdgeX-CLI-V2.md

```diff
@@ -0,0 +1,113 @@
+The latest version of edgex-cli  (1.0.1) only supports the V1 REST APIs and thus cannot be used with V2 releases of EdgeX.
+
+As the edgex-cli was developed organically over time, the current implementation has a number of bugs mostly involving a lack of consistent behavior, especially with respect to formatting of output.
+
+Other issues with the existing client include:
+
+- lack of tab completion
+- default output of commands is too verbose
+- verbose output sometime prevents use of jq 
+- static configuration file required (i.e. no registry support)
+- project hierarchy not conforming to best practice guidelines 
+
+## History
+The original Hanoi V1 client was created by a team at VMWare which is no longer participating in the project. 
+Canonical will lead the development of the Ireland/Jakarta V2 client.
+
+## Decision
+
+1. Use standardized command-line args/flags
+
+| Argument/Flag      | Description |
+| ----------- | ----------- |
+| `-d`, `--debug`      | show additional output for debugging purposes (e.g. REST URL, request JSON, …). This command-line arg will replace -v, --verbose and will no longer trigger output of the response JSON (see -j, --json).       |
+| `-j`, `--json`   | output the raw JSON response returned by the EdgeX REST API and *nothing* else. This output mode is used for script-based usage of the client.    |
+| `--version`   | output the version of the client and if available, the version of EdgeX installed on the system (using the version of the core data service)   |
+
+
+2. Restructure the Go code hierarchy to follow the [most recent recommended guidelines](https://github.com/golang-standards/project-layout). For instance /cmd should just contain the main application for the project, not an implementation for each command - that should be in /internal/cmd
+
+3. Take full advantage of the features of the underlying command-line library, [Cobra](https://github.com/spf13/cobra), such as tab-completion of commands.
+
+4. Change the client binary name from `edgex-cli` to `edgex`. The reason is that the "-cli" is superfluous. There is no `edgex` tool and best practice is to choose a simple, memorable word. CLI tool names do not generally include `"-cli"` in the name. (Examples: `go`,`snap`,`kubectl`,`git`,`docker` etc)
+
+5. Add support for the four common endpoints - `/ping`, `/metrics`, `/version`, `/status`
+
+6. Change the syntax of the CLI to require a positional service arg which precedes the command, thus allowing overlap of command names across services. The service name can either be an abbreviation or the full name of the service . Examples:
+
+    - `edgex data ping`
+    - `edgex d ping`
+    - `edgex metadata version`
+    - `edgex m version`
+    - `edgex metadata device list`
+    - `edgex command device list`
+
+    An alternative solution is to use a positional argument only for the common endpoints but specify it after the command or to have a regular arg like `-s`/`--service` `<name>`).
```

Comment:
This was discussed before with the VMWare team.  At the time, it was decided that the service was implicit in many commands because the request was only germane to a service.  
Would it be possible to do a little of each?  For example, if I did a request for events, that would not require the service name (either preceeding or post with a -s) because the service would be known to be associated to core data.  But if I asked for a device list, then the CLI would come back with an error indicating multiple services have that command and a more explicit service name is required.  Service name could always be provided from the beginning for extra clarity.  As to whether it is positional or after a -s or --service, I am for the latter.  Position arguments always confuse me when other parameters are also positional.

Reply:
Thanks for the comments. Yes, we discussed this in the Core WG meeting a week ago and I would tend to now agree what you both suggest, i.e. having a -s/--service for the cases where there is more than one possible service that provides the endpoint, but otherwise the CLI would be smart enough to know which one is being used.

Reply:
I have implemented this in [this PR](https://github.com/edgexfoundry/edgex-cli/pull/413) as suggested in your comments. For the common endpoints, it's now possible to specify the service to use: `-m/--metadata`, `-c/--command`, `-n/--notification`, `-s/--scheduler` or `--data` (which is the default). I updated this point in the ADR accordingly.